### PR TITLE
Add Support for Q_ENUM_NS and Q_NAMESPACE

### PR DESCRIPTION
**Pull request by LeonMatthesKDAB**

### CODE REVIEW

File: crates/cxx-qt-gen/src/generator/rust/qenum.rs [hunk 1]
```diff
@@ -4,29 +4,43 @@
...
+    qenums
+        .iter()
+        .flat_map(|qenum| {
+            let qenum_item = &qenum.item;
+            let qenum_ident = &qenum.ident;
+            let namespace = &qenum.namespace;
+            let namespace = if namespace.is_empty() {
+                quote! {}
+            } else {
+                quote! { #[namespace= #namespace ] }
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/generator/rust/qenum.rs:20]: add a space before the `=`

File: crates/cxx-qt-gen/src/parser/cxxqtdata.rs [hunk 2]
```diff
@@ -43,6 +50,8 @@ impl ParsedCxxQtData {
         Self {
             cxx_mappings: ParsedCxxMappings::default(),
             qobjects: BTreeMap::<Ident, ParsedQObject>::default(),
+            qenums: Vec::default(),
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/cxxqtdata.rs:53]: why `Vec::default()` vs just `vec![]` like elsewhere?

Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/parser/cxxqtdata.rs:53]: same for two lines below

File: crates/cxx-qt-gen/src/parser/qenum.rs [hunk 3]
```diff
@@ -54,12 +56,20 @@ impl ParsedQEnum {
...
 
+        let namespace = attribute_find_path(&qenum.attrs, &["namespace"])
+            .map(|attr_index| {
+                let attr = &qenum.attrs[attr_index];
+                expr_to_string(&attr.meta.require_name_value()?.value)
+            })
+            .transpose()?
+            .unwrap_or_default();
+
         // TODO: Add support for `namespace`, `cxx_name` and `rust_name` attributes.
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/qenum.rs:67]: namespace has been added now? so can be removed from the comment

File: crates/cxx-qt-gen/src/parser/qenum.rs [hunk 4]
```diff
@@ -96,18 +118,33 @@ mod tests {
                 Variant2,
             }
         };
+        let qenum: ItemEnum = syn::parse2(original_item.clone()).unwrap();
 
         let parsed = ParsedQEnum::parse(qenum).unwrap();
         assert_eq!(parsed.ident, "MyEnum");
+        assert_eq!(parsed.namespace, "");
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/qenum.rs:125]: if there is no namespace or QObject we should error ?

File: crates/cxx-qt-gen/src/parser/qnamespace.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    /// The name of the namespace
+    pub namespace: String,
+    /// whether qml_element was specified
+    pub qml_element: bool,
+}
+
+impl ParsedQNamespace {
+    pub fn parse(mut mac: ItemMacro) -> Result<Self> {
+        let namespace: LitStr = syn::parse2(mac.mac.tokens)?;
+        let namespace = namespace.value();
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/qnamespace.rs:20]: we should check here that the namespace is not empty?

File: crates/cxx-qt-gen/src/parser/qnamespace.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        Ok(Self {
+            namespace,
+            qml_element,
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::assert_eq;
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/qnamespace.rs:47]: why does this need to be included ? (and in `crates/cxx-qt-gen/src/generator/cpp/constructor.rs` in `main`) as elsewhere it doesn't need to be included ?

File: crates/cxx-qt-gen/src/generator/rust/qenum.rs (hunk 1 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/generator/rust/qenum.rs:20]: Done

File: crates/cxx-qt-gen/src/parser/cxxqtdata.rs (hunk 2 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/parser/cxxqtdata.rs:53]: No reason, really... can change.

File: crates/cxx-qt-gen/src/parser/qenum.rs (hunk 3 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/parser/qenum.rs:67]: Done

File: crates/cxx-qt-gen/src/parser/qnamespace.rs (hunk 5 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/parser/qnamespace.rs:20]: That's not done here.
The QEnum parser doesn't care whether the QEnum is assigned to anything, it's still a valid `ParsedQEnum`.

CxxQtData is responsible for matching the QEnum to the corresponding namespace or QObject.
Which will then error out if there's no namespace or QObject to associate the QEnum with.

File: crates/cxx-qt-gen/src/parser/qenum.rs (hunk 4 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/parser/qenum.rs:125]: That's not done here.
The QEnum parser doesn't care whether the QEnum is assigned to anything, it's still a valid `ParsedQEnum`.

CxxQtData is responsible for matching the QEnum to the corresponding namespace or QObject.
Which will then error out if there's no namespace or QObject to associate the QEnum with.

File: crates/cxx-qt-gen/src/parser/qnamespace.rs (hunk 5 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/parser/qnamespace.rs:20]: I've added a check that the namespace contains no whitespace and isn't empty.
We may want to do more checks in future, but this should be good enough to cover most cases.

File: crates/cxx-qt-gen/src/parser/qnamespace.rs (hunk 6 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/parser/qnamespace.rs:47]: it does, in fact, not need to be included :sweat_smile:

**Review by ahayzen-kdab — COMMENTED** [review]
Looks good, was there any reason for this still being a draft? And it needs rebasing...

**Review by ahayzen-kdab — APPROVED** [review]
Lets go with this, the book is being rewritten so docs can come later. And any other fixes can arrive later.

**Pull request merged**